pkg/controller/controller_ref_manager.go

```diff
@@ -0,0 +1,296 @@
+//       sense to check CanAdopt() again (e.g. in a different sync pass).
+func NewVirtualMachineControllerRefManager(
+	virtualMachineControl VirtualMachineControlInterface,
+	controller metav1.Object,
+	selector labels.Selector,
+	controllerKind schema.GroupVersionKind,
+	canAdopt func() error,
+) *VirtualMachineControllerRefManager {
+	return &VirtualMachineControllerRefManager{
+		BaseControllerRefManager: BaseControllerRefManager{
+			Controller:   controller,
+			Selector:     selector,
+			CanAdoptFunc: canAdopt,
+		},
+		controllerKind:        controllerKind,
+		virtualMachineControl: virtualMachineControl,
+	}
+}
+
+// ClaimVirtualMachines tries to take ownership of a list of VirtualMachines.
+//
+// It will reconcile the following:
+//   * Adopt orphans if the selector matches.
+//   * Release owned objects if the selector no longer matches.
+//
+// Optional: If one or more filters are specified, a VirtualMachine will only be claimed if
+// all filters return true.
+//
+// A non-nil error is returned if some form of reconciliation was attempted and
+// failed. Usually, controllers should try again later in case reconciliation
+// is still needed.
+//
+// If the error is nil, either the reconciliation succeeded, or no
+// reconciliation was necessary. The list of VirtualMachines that you now own is returned.
+func (m *VirtualMachineControllerRefManager) ClaimVirtualMachines(vms []*virtv1.VirtualMachine, filters ...func(machine *virtv1.VirtualMachine) bool) ([]*virtv1.VirtualMachine, error) {
+	var claimed []*virtv1.VirtualMachine
+	var errlist []error
+
+	match := func(obj metav1.Object) bool {
+		vm := obj.(*virtv1.VirtualMachine)
+		// Check selector first so filters only run on potentially matching VirtualMachines.
+		if !m.Selector.Matches(labels.Set(vm.Labels)) {
+			return false
+		}
+		for _, filter := range filters {
+			if !filter(vm) {
+				return false
+			}
+		}
+		return true
+	}
+	adopt := func(obj metav1.Object) error {
+		return m.AdoptVirtualMachine(obj.(*virtv1.VirtualMachine))
+	}
+	release := func(obj metav1.Object) error {
+		return m.ReleaseVirtualMachine(obj.(*virtv1.VirtualMachine))
+	}
+
+	for _, vm := range vms {
+		ok, err := m.ClaimObject(vm, match, adopt, release)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
+		if ok {
+			claimed = append(claimed, vm)
+		}
+	}
+	return claimed, utilerrors.NewAggregate(errlist)
+}
+
+// AdoptVirtualMachine sends a patch to take control of the vm. It returns the error if
+// the patching fails.
+func (m *VirtualMachineControllerRefManager) AdoptVirtualMachine(vm *virtv1.VirtualMachine) error {
+	if err := m.CanAdopt(); err != nil {
+		return fmt.Errorf("can't adopt VirtualMachine %v/%v (%v): %v", vm.Namespace, vm.Name, vm.UID, err)
+	}
+	// Note that ValidateOwnerReferences() will reject this patch if another
+	// OwnerReference exists with controller=true.
+	addControllerPatch := fmt.Sprintf(
```

Comment:
They are in general much cheaper. I wanted to keep the code as close as possible to the PodControllerRefManager. There a strategic merge patch is used (patches are much cheaper in general). I had to change it to a merge patch, because of the same reason why the gc does not work: https://github.com/kubernetes/kubernetes/issues/56348.

The PodControllerRefManager is here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_ref_manager.go#L211